### PR TITLE
Add `ocdev link` command

### PR DESCRIPTION
**Pull request by containscafeine**
This commit adds the `ocdev link` command which injects the
connection information from the target component to the source
component

### CODE REVIEW

**Comment by containscafeine** [comment]
Fix #342

File: pkg/component/component.go [hunk 1]
```diff
@@ -360,3 +362,91 @@ func Exists(client *occlient.Client, componentName string) (bool, error) {
 	}
 	return false, nil
 }
+
+// GetComponentService returns the Service object associated with the given
+// component.
+// An error is thrown when exactly one Service is not found for the component.
+func GetComponentService(client *occlient.Client, componentName string, applicationName string) (*corev1.Service, error) {
```
Comment (kadel) [inline pkg/component/component.go]: I'm not sure that this function belongs here.  If you really want to keep it here than it definitely shouldn't be public. Otherwise, it is a potential risk for creating cyclic imports like in the `storage` :-(

It might be better to do it similarly like here: https://github.com/redhat-developer/odo/pull/352/commits/f2d5b10ca01fec612f7349e606ba3966ea107574#diff-6025e9b620e9321c03c11bf29e5c4dbdR1190

File: pkg/component/component.go [hunk 2]
```diff
@@ -360,3 +362,91 @@ func Exists(client *occlient.Client, componentName string) (bool, error) {
...
+		return nil, fmt.Errorf("multiple Services exist for the selector: %v. Only one must be present", selector)
+	}
+
+	return &services[0], nil
+}
+
+// LinkInfo contains the information about the link being created between
+// components
+type LinkInfo struct {
+	ComponentName string
```
Comment (kadel) [inline pkg/component/component.go]: Is this source or target component name? 
Why is there only one?
There should be either comment, or better variable name should specify that.

File: pkg/component/component.go [hunk 3]
```diff
@@ -360,3 +362,91 @@ func Exists(client *occlient.Client, componentName string) (bool, error) {
...
+			{
+				Name:  linkPortKey,
+				Value: linkPortValue,
+			},
+		}, sourceDC); err != nil {
+		return nil, errors.Wrapf(err, "unable to add environment variables to Deployment Config %v", sourceDC.Name)
+	}
+
+	return &LinkInfo{
+		ComponentName: sourceComponent,
```
Comment (kadel) [inline pkg/component/component.go]: Why are we returning just sourceComponent?

File: pkg/component/component.go (hunk 1 above)
Comment (containscafeine) [reply pkg/component/component.go]: moved this similar to the current deployment config logic

File: pkg/component/component.go (hunk 2 above)
Comment (containscafeine) [reply pkg/component/component.go]: added source and target components

File: pkg/component/component.go (hunk 3 above)
Comment (containscafeine) [reply pkg/component/component.go]: modified LinkInfo to include both source and target components

File: cmd/link.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	Short: "Link target component to source component",
+	Long: `Link target component to source component
+If source component is not provided, the link is created to the current active
+component.
+In the linking process, the environment variables containing the connection
+information from target component are injected into the source component and
+printed to STDOUT.
+`,
+	Example: `
+	# Link current component to a component 'nodejs'
```
Comment (kadel) [inline cmd/link.go]: There should be just two spaces at the beginning of each line, to make it properly align

File: cmd/link.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+	ocdev link nodejs
```
Comment (kadel) [inline cmd/link.go]: Better example would be to do `ocdev link mariadb` as you will be injecting variables about mariadb  to current component

File: cmd/link.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+	# Link 'nodejs' component to 'mariadb' component
+	ocdev link nodejs --component mariadb
```
Comment (kadel) [inline cmd/link.go]: again this should be the other way around `ocdev link mariadb --component nodejs`

File: cmd/link.go (hunk 4 above)
Comment (containscafeine) [reply cmd/link.go]: fixed

File: cmd/link.go (hunk 5 above)
Comment (containscafeine) [reply cmd/link.go]: fixed

File: cmd/link.go (hunk 6 above)
Comment (containscafeine) [reply cmd/link.go]: fixed

File: cmd/util.go [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+// getComponent returns the component to be used for the operation. If an input
+// component is specified, then it is returned, if not, the current component
+// is fetched and returned
+func getComponent(client *occlient.Client, inputComponent, applicationName, projectName string) string {
+	if len(inputComponent) == 0 {
+		c, err := component.GetCurrent(client, applicationName, projectName)
+		checkError(err, "Could not get current component")
+		return c
+	}
+	return inputComponent
```
Comment (kadel) [inline cmd/util.go]: It would be good to check if the component exists here, so we don't have to do it each time we call this function.

(we need to check it only if `len(inputComponent)!=0`

File: pkg/component/component.go [hunk 8]
```diff
@@ -7,11 +7,15 @@ import (
 	buildv1 "github.com/openshift/api/build/v1"
 	"github.com/pkg/errors"
 	log "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+
+	"strings"
```
Comment (kadel) [inline pkg/component/component.go]: please group this with other standard library imports

File: pkg/component/component.go [hunk 9]
```diff
@@ -239,3 +243,83 @@ func Exists(client *occlient.Client, componentName, applicationName, projectName
...
+	SourceComponent string
+	TargetComponent string
+	Envs            []string
+}
+
+// Link injects connection information of the target component into the source
+// component as environment variables
+func Link(client *occlient.Client, sourceComponent, targetComponent, applicationName string) (*LinkInfo, error) {
+	// Get Service associated with the target component
+	targetService, err := client.GetOneServiceFromSelector(
```
Comment (kadel) [inline pkg/component/component.go]: my head hurts from trying to parse this :-( 
Would it be better to have separate this? It is hard to parse 3 nested function calls :(

File: cmd/util.go (hunk 7 above)
Comment (containscafeine) [reply cmd/util.go]: added

File: pkg/component/component.go (hunk 8 above)
Comment (containscafeine) [reply pkg/component/component.go]: fixed

File: pkg/component/component.go [hunk 10]
```diff
@@ -239,3 +243,83 @@ func Exists(client *occlient.Client, componentName, applicationName, projectName
...
+		TargetComponent: targetComponent,
+		Envs: []string{
+			fmt.Sprintf("%v=%v", linkHostKey, linkHostValue),
+			fmt.Sprintf("%v=%v", linkPortKey, linkPortValue),
+		},
+	}, nil
+}
+
+// getPortFromService returns the first port listed in the service
+func getPortFromService(service *corev1.Service) (int32, error) {
```
Comment (kadel) [inline pkg/component/component.go]: Do we really need a separate function that is just one if statement and nothing more? This is used just in one place.

File: cmd/util.go [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+// getComponent returns the component to be used for the operation. If an input
+// component is specified, then it is returned, if not, the current component
+// is fetched and returned
+func getComponent(client *occlient.Client, inputComponent, applicationName, projectName string) string {
+	if len(inputComponent) == 0 {
+		c, err := component.GetCurrent(client, applicationName, projectName)
+		checkError(err, "Could not get current component")
+		return c
+	}
+	exists, err := component.Exists(client, inputComponent, applicationName, projectName)
```
Comment (kadel) [inline cmd/util.go]: Do we need to check that component exists if we got it from `GetCurrent`?
`Exists` is expensive it might be enough to put into `else` block for previous if.

File: pkg/component/component.go (hunk 9 above)
Comment (containscafeine) [reply pkg/component/component.go]: haha, fixed

File: pkg/component/component.go (hunk 10 above)
Comment (containscafeine) [reply pkg/component/component.go]: imo it makes the Link() function look cleaner

File: cmd/util.go (hunk 11 above)
Comment (containscafeine) [reply cmd/util.go]: we're not checking if we're getting it from GetCurrent. The if block has a return statement, we don't need an else here

File: pkg/component/component.go (hunk 10 above)
Comment (kadel) [reply pkg/component/component.go]: IMO it makes it harder to read ;-)

File: cmd/util.go (hunk 11 above)
Comment (kadel) [reply cmd/util.go]: ah, sorry, Didn't notice return there. LGTM than ;-)
